extractor.go

```diff
@@ -186,87 +188,112 @@ func (e *Extractor) Tar(ctx context.Context, body io.Reader, location string, re
 
 // Zip extracts a .zip archived stream of data in the specified location.
 // It accepts a rename function to handle the names of the files (see the example).
-func (e *Extractor) Zip(ctx context.Context, body io.Reader, location string, rename Renamer) error {
+func (e *Extractor) Zip(originalCtx context.Context, body io.Reader, location string, rename Renamer) error {
 	// read the whole body into a buffer. Not sure this is the best way to do it
 	buffer := bytes.NewBuffer([]byte{})
-	copyCancel(ctx, buffer, body)
+	copyCancel(originalCtx, buffer, body)
 
 	archive, err := zip.NewReader(bytes.NewReader(buffer.Bytes()), int64(buffer.Len()))
 	if err != nil {
 		return errors.Annotatef(err, "Read the zip file")
 	}
 
-	links := []link{}
-
-	// We make the first pass creating the directory structure, or we could end up
-	// attempting to create a file where there's no folder
-	for _, header := range archive.File {
-		select {
-		case <-ctx.Done():
-			return errors.New("interrupted")
-		default:
+	fileChan := make(chan *zip.File)
+	go func() {
+		defer close(fileChan)
+		for _, header := range archive.File {
+			fileChan <- header
 		}
+	}()
 
-		path := header.Name
+	links := []link{}
+	g, ctx := errgroup.WithContext(originalCtx)
+
+	filesWorker := func() error {
+		for header := range fileChan {
```

Comment:
Can we move `fileWorker` out and transform into a normal function?

Based on my experience, I'd avoid mixing "anonymous functions" and "goroutines", because in golang anonymous functions have access to the local scope of the containing-function and do not create closure. This is very counterintuitive and risky because we need to synchronize access to local variables declared in the container function

For example:
https://github.com/codeclysm/extract/pull/16/files?diff=unified&w=1#diff-705c94be3d96b8c2131b6430f14904b30b254b30ca4ce3229aecd7db83822b15L246

 in this case inside the goroutine we append to the `links` array without synchronization, but this is wrong because it is a shared array declared outside the goroutine.

Reply:
I don't think it would be possible, we can't pass arguments to the goroutine when launched with `errorgroup.Go()`. How would do it? :thinking: 

The append to `links` is certainly an issue though, I'll find a solution.

Reply:
mmm, interesting, I don't know, let me try something